pkg/kubecost/allocation.go

```diff
@@ -2431,12 +2494,26 @@ func (asr *AllocationSetRange) UTCOffset() time.Duration {
 // Window returns the full window that the AllocationSetRange spans, from the
 // start of the first AllocationSet to the end of the last one.
 func (asr *AllocationSetRange) Window() Window {
-	if asr == nil || asr.Length() == 0 {
+	asr.Lock()
+	defer asr.Unlock()
+	return asr.window()
+}
+
+func (asr *AllocationSetRange) window() Window {
+	var length int
+
+	if asr == nil || asr.allocations == nil {
+		length = 0
+	} else {
+		length = len(asr.allocations)
+	}
+
+	if asr == nil || length == 0 {
```

Comment:
Go convention generally asks you to return early rather than do this sort of if statement followed by a later check and return. I think the logic that was present beforehand adhered to this, like so:

```go
if asr == nil || asr.allocations == nil {
    return NewWindow(nil, nil)
}
```

There's also decent amount of repetition here that I'm not a huge fan of. Why not just leave the original logic as it was?

Reply:
Yeah this was the section I mentioned that I really didn't like either. I did manage to find a good work around by calling asr.Window() before locking asr and can remove all of this. Let me know if this work around has any inherent issues

Reply:
Specifically this line:

https://github.com/kubecost/cost-model/pull/1040/files#diff-7c339daaa6b130f0e74be88b76e4c35d3941edf78f886fae793772b8f566c58cR2269

Reply:
> However does this have the potential of asr data changing in between unlocking and locking again even if in succession?

Yes, I prefer your initial private, non-locking, non-safe `.window()` over trying to lock twice in `AccumulateBy`.

Reply:
No worries and this makes sense to me, went ahead and reverted and improved. Thanks!